File: Source/app/src/main/java/com/amgregoire/mangafeed/UI/Fragments/MangaInfoFragment.java
package com.amgregoire.mangafeed.UI.Fragments;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.amgregoire.mangafeed.Common.MangaEnums;
import com.amgregoire.mangafeed.MangaFeed;
import com.amgregoire.mangafeed.Models.Manga;
import com.amgregoire.mangafeed.R;
import com.amgregoire.mangafeed.UI.Activities.NavigationActivity;
import com.amgregoire.mangafeed.UI.Adapters.MangaInfoChaptersAdapter;
import com.amgregoire.mangafeed.UI.BackHandledFragment;
import com.amgregoire.mangafeed.UI.Mappers.IManga;
import com.amgregoire.mangafeed.UI.Presenters.MangaInfoPres;
import com.amgregoire.mangafeed.Utils.DownloadManager;
import com.amgregoire.mangafeed.Utils.MangaLogger;
import com.l4digital.fastscroll.FastScrollRecyclerView;

import butterknife.BindColor;
import butterknife.BindDrawable;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Andy Gregoire on 3/12/2018.
 */

public class MangaInfoFragment extends BackHandledFragment implements IManga.MangaMap
{
    public final static String TAG = MangaInfoFragment.class.getSimpleName();
    public final static String MANGA_KEY = TAG + "MANGA";
    public final static String OFFLINE_KEY = TAG + "OFFLINE";

    @BindView(R.id.recyclerViewMangaInfo) FastScrollRecyclerView mRecyclerView;
    @BindView(R.id.bottomNavigationMangaInfo) BottomNavigationView mBottomNav;
    @BindView(R.id.swipeRefreshLayoutMangaInfo) SwipeRefreshLayout mSwipeLayout;
    @BindView(R.id.toolbar) Toolbar mToolbar;

    @BindColor(R.color.manga_blue) int mColorBlue;
    @BindColor(R.color.manga_red) int mColorRed;
    @BindColor(R.color.manga_green) int mColorGreen;
    @BindColor(R.color.manga_gray) int mColorGray;
    @BindColor(R.color.manga_white) int mColorWhite;

    @BindDrawable(R.drawable.ic_check_circle_outline_white_24dp) Drawable mSelectFilled;
    @BindDrawable(R.drawable.ic_checkbox_blank_circle_outline_white_24dp) Drawable mSelectOutline;

    private IManga.MangaPres mPresenter;

    public static Fragment newInstance(Manga manga, boolean offline)
    {
        Bundle lBundle = new Bundle();
        lBundle.putParcelable(MANGA_KEY, manga);
        lBundle.putBoolean(OFFLINE_KEY, offline);

        Fragment lFragment = new MangaInfoFragment();
        lFragment.setArguments(lBundle);

        return lFragment;
    }

    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState)
    {
        setHasOptionsMenu(true);

        View lView = inflater.inflate(R.layout.fragment_manga_info, null);
        ButterKnife.bind(this, lView);

        mPresenter = new MangaInfoPres(this);
        mPresenter.init(getArguments());

        return lView;
    }

    @Override
    public void onResume()
    {
        super.onResume();
        mPresenter.subEventBus();
    }

    @Override
    public void onPause()
    {
        super.onPause();
        mPresenter.unSubEventBus();
    }

    @Override
    public void initViews()
    {
        setupBottomNav();
        ((NavigationActivity) getActivity()).setSupportActionBar(mToolbar);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater)
    {
        if (!mPresenter.isDownload())
        {
            if (!mPresenter.isOffline())
            {
                inflater.inflate(R.menu.menu_toolbar_home_manga_info, menu);
            }
            else
            {
                inflater.inflate(R.menu.menu_toolbar_downloads_manga_info, menu);
            }

            mToolbar.setNavigationIcon(R.drawable.navigation_back);
            mToolbar.setTitle(mPresenter.getTitle());
        }
        else
        {
            if (!mPresenter.isOffline())
            {
                inflater.inflate(R.menu.menu_toolbar_home_manga_info_download, menu);
            }
            else
            {
                inflater.inflate(R.menu.menu_toolbar_downloads_manga_info_remove, menu);
            }

            mToolbar.setNavigationIcon(R.drawable.ic_checkbox_blank_circle_outline_white_24dp);
            mToolbar.setTitle("Select Items");
            onDownloadViewEnabled();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        switch (item.getItemId())
        {
            case R.id.menuMangaInfoRefresh:
                onRefreshInfo();
                break;
            case R.id.menuMangaInfoDownload:
                toggleDownloadingFlag();
                onDownloadViewEnabled();
                break;
            case R.id.menuDownloadsMangaInfoDownload:
                toggleDownloadingFlag();
                getActivity().invalidateOptionsMenu();
                onDownloadViewEnabled();
                break;
            case R.id.menuDownloadMangaInfoCancel:
                onDownloadCancel();
                break;
            case R.id.menuMangaInfoDownloadDownload:
                if (DownloadManager.isStoragePermissionGranted(getActivity()))
                {
                    onDownloadDownload(); // start download
                    onDownloadCancel(); // exit download view
                    MangaFeed.getInstance().makeToastShort("Starting downloads now");
                }
                else
                {
                    MangaFeed.getInstance()
                             .makeToastShort("Need storage permissions to download chapters.");
                }
                break;
            case R.id.menuDownloadsMangaInfoRemove:
                onDownloadRemove();
                break;
            case android.R.id.home:
                if (mPresenter.isDownload())
                {
                    if (mToolbar.getNavigationIcon() == mSelectFilled)
                    {
                        mToolbar.setNavigationIcon(mSelectOutline);
                        onSelectAllOrNone(false);
                    }
                    else
                    {
                        mToolbar.setNavigationIcon(mSelectFilled);
                        onSelectAllOrNone(true);
                    }
                }
                else
                {
                    getActivity().onBackPressed();
                }
                break;
            default:
                return super.onOptionsItemSelected(item);
        }

        return true;
    }

    @Override
    public void setBottomNavStartContinue(String readText)
    {
        mBottomNav.getMenu().findItem(R.id.menuMangaInfoBottomContinueReading).setTitle(readText);
    }

    @Override
    public void setBottomNavFollowTitle(int followType)
    {
        MenuItem item = mBottomNav.getMenu().findItem(R.id.menuMangaInfoBottomNavFollow);
        if (followType == 0)
        {
            item.setIcon(R.drawable.ic_heart_outline_white_24dp);
        }
        else
        {
            item.setIcon(R.drawable.ic_heart_white_24dp);
        }

        item.setTitle(MangaEnums.FollowType.values()[followType].toString());
    }

    @Override
    public void toggleDownloadingFlag()
    {
        mPresenter.toggleDownload();
        getActivity().invalidateOptionsMenu();
    }

    @Override
    public void registerAdapter(RecyclerView.Adapter adapter, RecyclerView.LayoutManager manager)
    {
        mRecyclerView.setLayoutManager(manager);
        mRecyclerView.setAdapter(adapter);
    }

    @Override
    public void startRefresh()
    {
        mSwipeLayout.setEnabled(true);
        mSwipeLayout.setRefreshing(true);
    }

    @Override
    public void stopRefresh()
    {
        mSwipeLayout.setRefreshing(false);
        mSwipeLayout.setEnabled(false);
    }

    @Override
    public String getTagText()
    {
        return TAG;
    }

    @Override
    public boolean onBackPressed()
    {
        if (mPresenter.isDownload())
        {
            onDownloadCancel();
            return false;
        }

        return true;
    }

    /***
     * This function tells the presenter the (un)select all button has been selected.
     *
     * @param isAll
     */
    public void onSelectAllOrNone(boolean isAll)
    {
        mPresenter.onSelectAllOrNone(isAll);
    }

    /***
     * This function makes the recyclerview scroll to the selected position when the download view is enabled.
     *
     */
    public void onDownloadViewStart()
    {
        int lPosition = ((MangaInfoChaptersAdapter) mRecyclerView.getAdapter()).getFirstDownloadScrollPosition();
        mRecyclerView.scrollToPosition(lPosition);
    }

    /***
     * This function tells the presenter the cancel button has been selected.
     *
     */
    public void onDownloadCancel()
    {
        mPresenter.onDownloadCancel();
        mBottomNav.setVisibility(View.VISIBLE);
    }

    /***
     * This function tells the presenter the download button has been selected.
     *
     */
    public void onDownloadDownload()
    {
        mPresenter.onDownloadDownload();
    }

    /***
     * This function tells the presenter the remove menu button has been selected.
     *
     */
    public void onDownloadRemove()
    {
        mPresenter.onDownloadRemove();
    }

    /***
     * This function tells the presenter to refresh the information view.
     *
     */
    public void onRefreshInfo()
    {
        mPresenter.onRefreshInfo();
    }

    /***
     * This function tells the presenter to enable Download view.
     *
     */
    public void onDownloadViewEnabled()
    {
        mPresenter.onDownloadViewEnabled();
        mBottomNav.setVisibility(View.GONE);
        onDownloadViewStart();
    }

    /***
     * This function sets up the bottom navigation view interactions.
     *
     */
    private void setupBottomNav()
    {
        mBottomNav.setOnNavigationItemSelectedListener((MenuItem item) ->
        {
            switch (item.getItemId())
            {
                case R.id.menuMangaInfoBottomNavFollow:
                    // pop up menu
                    try
                    {
                        PopupMenu lPopupMenu = new PopupMenu(getContext(), mBottomNav);
                        lPopupMenu.getMenuInflater()
                                  .inflate(R.menu.menu_follow_status, lPopupMenu.getMenu());

                        lPopupMenu.setOnMenuItemClickListener(popupItem ->
                        {
                            int lStatus;

                            switch (popupItem.getItemId())
                            {
                                case R.id.menuFollowStatusReading:
                                    lStatus = Manga.FOLLOW_READING;
                                    break;
                                case R.id.menuFollowStatusCompleted:
                                    lStatus = Manga.FOLLOW_COMPLETE;
                                    break;
                                case R.id.menuFollowStatusOnHold:
                                    lStatus = Manga.FOLLOW_ON_HOLD;
                                    break;
                                case R.id.menuFollowStatusPlanToRead:
                                    lStatus = Manga.FOLLOW_PLAN_TO_READ;
                                    break;
                                default:
                                    lStatus = Manga.UNFOLLOW;
                                    break;
                            }

                            mPresenter.onUpdateFollowStatus(lStatus);

                            return true;
                        });

                        lPopupMenu.show(); //showing lPopupMenu menu
                    }
                    catch (NullPointerException npe)
                    {
                        MangaLogger.logError(TAG, "Context was null", npe.getMessage());
                    }
                    return true;
                case R.id.menuMangaInfoBottomContinueReading:
                    // Start Reading from latest chapter
                    return true;
            }
            return false;
        });
    }
}

File: Source/app/src/main/java/com/amgregoire/mangafeed/v2/ui/info/MangaInfoFragment.kt
package com.amgregoire.mangafeed.v2.ui.info

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.PopupMenu
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.amgregoire.mangafeed.MangaFeed
import com.amgregoire.mangafeed.R
import com.amgregoire.mangafeed.uiScope
import com.amgregoire.mangafeed.v2.NavigationType
import com.amgregoire.mangafeed.v2.ResourceFactory
import com.amgregoire.mangafeed.v2.enums.FollowType
import com.amgregoire.mangafeed.v2.repository.local.LocalMangaRepository
import com.amgregoire.mangafeed.v2.ui.base.BaseFragment
import com.amgregoire.mangafeed.v2.ui.base.FragmentNavMap
import com.amgregoire.mangafeed.v2.ui.map.ToolbarMap
import com.amgregoire.mangafeed.v2.ui.read.ReaderFragment
import com.amgregoire.mangafeed.v2.ui.read.ReaderViewModel
import kotlinx.android.synthetic.main.fragment_manga_info2.view.*
import kotlinx.coroutines.launch


/**
 * Created by Andy Gregoire on 3/12/2018.
 */

class MangaInfoFragment : BaseFragment()
{
    private val localMangaRepository = LocalMangaRepository()

    private val readerViewModel: ReaderViewModel? by lazy {
        val parent = activity ?: return@lazy null
        ViewModelProviders.of(parent).get(ReaderViewModel::class.java)
    }

    private val mangaInfoViewModel by lazy {
        val manga = localMangaRepository.getManga(arguments!!.getString(MANGA_LINK_KEY), arguments!!.getString(MANGA_SOURCE_KEY)) ?: return@lazy null
        ViewModelProviders.of(this, MangaInfoViewModelFactory(MangaFeed.app, manga!!)).get(MangaInfoViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
    {
        self = inflater.inflate(R.layout.fragment_manga_info2, null)
        return self
    }

    override fun onStart()
    {
        super.onStart()

        updateParentSettings()
        setupBottomNav()

        mangaInfoViewModel?.state?.observe(this, Observer { state ->
            when (state)
            {
                is MangaInfoViewModel.State.Complete -> uiScope.launch { renderComplete() }
                is MangaInfoViewModel.State.Loading -> uiScope.launch { renderLoading() }
                is MangaInfoViewModel.State.Failed -> uiScope.launch { renderFailed() }
            }
        })

        mangaInfoViewModel?.mangaInfo?.observe(this, Observer { mangaInfo ->
            mangaInfo ?: return@Observer

            if (self.rvMangaInfo.adapter != null)
            {
                (self.rvMangaInfo.adapter as MangaInfoAdapter).updateInfo(mangaInfo.manga, mangaInfo.dbChapters)
                return@Observer
            }

            mangaInfoViewModel?.setFollowStatus(mangaInfo.manga.followType)

            self.rvMangaInfo.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
            self.rvMangaInfo.adapter = MangaInfoAdapter(
                    manga = mangaInfo.manga,
                    source = MangaFeed.app.currentSource,
                    dbChapters = mangaInfo.dbChapters,
                    chapterSelected = { manga, chapters, chapter ->
                        val parent = activity ?: return@MangaInfoAdapter
                        readerViewModel?.updateReaderInfo(manga, chapters, chapter, true)
                        (parent as FragmentNavMap).addFragmentParent(ReaderFragment.newInstance(), this, ReaderFragment.TAG)
                    }
            )
        })

        mangaInfoViewModel?.mangaInfoBottomNav?.observe(this, Observer { bottomNavInfo ->
            bottomNavInfo ?: return@Observer

            val item = self.bottomNavMangaInfo.menu.findItem(R.id.menuMangaInfoBottomContinueReading)
            item.title = getString(bottomNavInfo.startText)

            val followItem = self.bottomNavMangaInfo.menu.findItem(R.id.menuMangaInfoBottomNavFollow)
            followItem.setIcon(bottomNavInfo.followIcon)
            followItem.title = getString(bottomNavInfo.followText)
        })

        self.swipeRefreshMangaInfo.setOnRefreshListener {
            self.swipeRefreshMangaInfo.isRefreshing = false
            mangaInfoViewModel?.refresh()
        }

        self.emptyStateMangaInfo.setButtonClickListener(View.OnClickListener {
            self.swipeRefreshMangaInfo.isRefreshing = false
            mangaInfoViewModel?.refresh()
        })
    }

    override fun updateParentSettings()
    {
        val parent = activity ?: return
        mangaInfoViewModel?.manga?.name?.let {  (parent as ToolbarMap).setTitle(it)}
        (parent as ToolbarMap).setNavigationIcon(ResourceFactory.getNavigationIcon(NavigationType.Close))
        (parent as ToolbarMap).setOptionsMenu(R.menu.menu_empty)
    }

    private fun renderLoading()
    {
        self.emptyStateMangaInfo.showLoader()
    }

    private fun renderComplete()
    {
        self.emptyStateMangaInfo.hide()
    }

    private fun renderFailed()
    {
        self.emptyStateMangaInfo.hideLoader(true)
    }

    /***
     * This function sets up the bottom navigation view interactions.
     *
     */
    private fun setupBottomNav()
    {
        self.bottomNavMangaInfo.setOnNavigationItemSelectedListener { item: MenuItem ->
            when (item.itemId)
            {
                R.id.menuMangaInfoBottomNavFollow ->
                {
                    val context = context ?: return@setOnNavigationItemSelectedListener false
                    val popMenu = PopupMenu(context, self.bottomNavMangaInfo)

                    popMenu.menuInflater.inflate(R.menu.menu_follow_status, popMenu.menu)

                    popMenu.setOnMenuItemClickListener { popupItem ->
                        when (popupItem.itemId)
                        {
                            R.id.menuFollowStatusReading -> mangaInfoViewModel?.setFollowStatus(FollowType.Reading)
                            R.id.menuFollowStatusCompleted -> mangaInfoViewModel?.setFollowStatus(FollowType.Completed)
                            R.id.menuFollowStatusOnHold -> mangaInfoViewModel?.setFollowStatus(FollowType.On_Hold)
                            R.id.menuFollowStatusPlanToRead -> mangaInfoViewModel?.setFollowStatus(FollowType.Plan_to_Read)
                            else -> mangaInfoViewModel?.setFollowStatus(FollowType.Unfollow)
                        }
                        true
                    }

                    popMenu.show()

                    return@setOnNavigationItemSelectedListener true
                }
                R.id.menuMangaInfoBottomContinueReading ->
                {
                    // Start Reading from latest chapter
                    val parent = activity ?: return@setOnNavigationItemSelectedListener true

                    val adapter = (self.rvMangaInfo.adapter as? MangaInfoAdapter) ?: return@setOnNavigationItemSelectedListener true
                    val manga = adapter.manga
                    val chapters = adapter.dbChapters
                    val chapter = adapter.dbChapters.firstOrNull { chapter -> chapter.url == manga.recentChapter } ?: chapters[chapters.size - 1]

                    readerViewModel?.updateReaderInfo(manga, chapters, chapter, true)

                    (parent as FragmentNavMap).addFragmentParent(ReaderFragment.newInstance(), this, ReaderFragment.TAG)
                    return@setOnNavigationItemSelectedListener true
                }
                else -> false
            }
        }
    }

    companion object
    {
        val TAG = MangaInfoFragment::class.java.simpleName
        val MANGA_LINK_KEY = TAG + "LINK"
        val MANGA_SOURCE_KEY = TAG + "SOURCE"
        val OFFLINE_KEY = TAG + "OFFLINE"

        fun newInstance(link: String, source: String, offline: Boolean): androidx.fragment.app.Fragment
        {
            val lBundle = Bundle()
            lBundle.putString(MANGA_LINK_KEY, link)
            lBundle.putString(MANGA_SOURCE_KEY, source)
            lBundle.putBoolean(OFFLINE_KEY, offline)

            val lFragment = MangaInfoFragment()
            lFragment.arguments = lBundle

            return lFragment
        }
    }
}
